src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -596,15 +577,54 @@ void GcsPlacementGroupScheduler::AcquireBundleResources(
   }
 }
 
+absl::flat_hash_map<scheduling::NodeID, ResourceRequest> ToNodeBundleResourcesMap(
+    const std::shared_ptr<BundleLocations> &bundle_locations) {
+  absl::flat_hash_map<scheduling::NodeID, ResourceRequest> node_bundle_resources_map;
+  for (auto &bundle : *bundle_locations) {
+    auto node_id = scheduling::NodeID(bundle.second.first.Binary());
+    const auto &bundle_spec = *bundle.second.second;
+    auto bundle_resource_request = ResourceMapToResourceRequest(
+        bundle_spec.GetFormattedResources(), /*requires_object_store_memory=*/false);
+    node_bundle_resources_map[node_id] += bundle_resource_request;
+  }
+  return node_bundle_resources_map;
+}
+
+void GcsPlacementGroupScheduler::CommitBundleResources(
+    const std::shared_ptr<BundleLocations> &bundle_locations) {
+  // Acquire bundle resources from gcs resources manager.
+  auto &cluster_resource_manager =
+      cluster_resource_scheduler_.GetClusterResourceManager();
+  auto node_bundle_resources_map = ToNodeBundleResourcesMap(bundle_locations);
+  for (const auto &[node_id, node_bundle_resources] : node_bundle_resources_map) {
+    for (const auto &[resource_id, capacity] : node_bundle_resources.ToMap()) {
+      cluster_resource_manager.UpdateResourceCapacity(
+          node_id, resource_id, capacity.Double());
+    }
+  }
+}
+
 void GcsPlacementGroupScheduler::ReturnBundleResources(
     const std::shared_ptr<BundleLocations> &bundle_locations) {
-  // Release bundle resources to gcs resources manager.
+  // Return bundle resources to gcs resources manager should contains the following steps.
+  // 1. Remove related bundle resources from nodes.
+  // 2. Add resources allocated for bundles back to nodes.
+  // TODO(Shanly): It's necessary to check whether the bundle could be returned in gcs
+  // scheduling. Only bundles with same total and availe resources could be returnd.
   auto &cluster_resource_manager =
       cluster_resource_scheduler_.GetClusterResourceManager();
   for (auto &bundle : *bundle_locations) {
+    auto node_id = scheduling::NodeID(bundle.second.first.Binary());
+    const auto &bundle_spec = *bundle.second.second;
+    // Remove bundle resource names (the label with `_group_`).
+    std::vector<scheduling::ResourceID> bundle_resource_ids;
+    for (const auto &entry : bundle_spec.GetFormattedResources()) {
+      bundle_resource_ids.emplace_back(scheduling::ResourceID(entry.first));
+    }
+    cluster_resource_manager.DeleteResources(node_id, bundle_resource_ids);
```

Comment:
is it no-op if the resource ids don't exist?

Reply:
I am asking it because there are situations this function is called when bundle resources are not created yet (e.g., prepare failure). Can you add a comment in these cases this will become a no-op?